Add fallback support to nodejs server side libraries

LibanOdowa opened the pull request:
The main point of this change was to allow clients to use the fallback support we added for browsers in nodeJS, allowing them to bypass gRPC if needed. This is implemented through a 'fallback' option in the GrpcClient class's options. This will need to later be added to client library generation but for an example of how this would be implemented, you can look at the echo_client.js under the tests.

Unified tests to use one version of echo client rather than two

alexander-fenster commented on browser-test/mock-server.js:
```diff
@@ -0,0 +1,83 @@
+/**
```
In #565 I moved it to `system-test/mock-server.js`, I believe this file is not needed anymore

alexander-fenster commented on package.json:
```diff
@@ -57,7 +57,7 @@
     "karma": "^4.1.0",
     "karma-child-process": "^1.0.3",
-    "karma-chrome-launcher": "^3.0.0",
+    "karma-chrome-launcher": "^2.2.0",
```
Please undo this change, it got updated by Renovate bot in #564 :)

alexander-fenster commented on src/browser.ts:
```diff
@@ -187,6 +188,8 @@ export class GrpcClient {
 
       const url = `${grpcFallbackProtocol}://${servicePath}:${servicePort}/$rpc/${protoServiceName}/${rpcName}`;
 
+      const nodeFetch = require('node-fetch');
```
since it's TypeScript, can we `import * as nodeFetch from 'node-fetch'` instead? (on top of the file)

Also, please make sure `node-fetch` is in `package.json`.

alexander-fenster replied:
Example of the similar code in another package we use: https://github.com/googleapis/gaxios/blob/19b4b3c7dffb0f01bd475932fab5cc5642a19c23/src/gaxios.ts#L16-L33

alexander-fenster commented on system-test/fixtures/google-gax-packaging-test-app/protos/protos.json:
```diff
@@ -0,0 +1,1281 @@
+{
```
Please remove `google-gax.tgz` from this change (the file above) - we don't need to commit it.

alexander-fenster commented on system-test/fixtures/google-gax-packaging-test-app/protos/protos.json:
```diff
@@ -0,0 +1,1281 @@
+{
+  "nested": {
```
Also, if we add `protos.json`, we might want to remove the old `echo.json`.

LibanOdowa replied:
Done. Got rid of the client as well.

alexander-fenster commented:
Also, CI fails because of `gts` - running `npx gts fix` should help.

alexander-fenster commented:
@ramya-ramalingam Could you please also take a look at the test changes?

ramya-ramalingam commented on system-test/fixtures/google-gax-packaging-test-app/src/index.js:
```diff
@@ -40,18 +40,38 @@ if (require.main === module) {
+  await testChat(grpcClient);
+  await testWait(grpcClient);
+
+  await testEcho(fallbackClient);
```
We might want to mention here that streaming calls aren't currently supported when using gRPC-fallback. Or alternately have tests making sure an exception is thrown when we try to use streaming calls.

alexander-fenster replied:
Making sure an exception is thrown is a great idea, since it (a) checks the intended behavior, and (b) explains itself clearly.  Since we don't use `chai` in this file, just regular 

```js
try {
  await stuff();
  throw new Error("stuff is supposed to throw because it's not supported");
}
catch (err) {
  // ignoring the error
}
```
should work and show the idea.

LibanOdowa replied:
Good idea, done.

alexander-fenster commented:
Okay, I just figured we might to do a little bit more stuff with auth to make it work in Node. Let's see what kind of auth parameters we can pass.

To make auth work, we should finally have an instance of `OAuth2Client` so that we could ask it for request headers (`client.getRequestHeaders`). How can we get this auth client instance?

1. Pure browser usage (`if (isBrowser())`): we need to have an `OAuth2Client` instance passed. We currently ask users to put it into `auth` parameter.

2. Node.js usage (`if (!isBrowser())`): we might have a service account key, so the regular auth stuff from Gax should work:

```js
this.auth = options.auth || new GoogleAuth(options); // reads service account key
const client = await this.auth.getClient();
```

There is a discrepancy here: for browser, we expect `auth` to be an authenticated instance of `OAuth2Client`, while for Node.js `auth` is supposed to be an instance of `GoogleAuth`.

To make these two use cases work in a similar fashion, let's do this: if we're not in browser, let's copy the logic from Node.js scenario. The code I'm thinking about would be: instead of using `this.auth.getRequestHeader()` (which adds to the ambiguity), let's call it `this.authClient`:

In constructor (which must be synchronous):
```js
if (isBrowser()) {
    if (!options.auth) {
      throw new Error(
        'You need to pass auth instance to use gRPC-fallback client in browser. Use OAuth2Client from google-auth-library.'   // note: text has changed a little bit
      );
    }
  this.authClient = options.auth;
} else {
  this.googleAuth = options.auth || new GoogleAuth(options);
  // note: this.authClient will be undefined since we can't do `await this.googleAuth.getClient()` in the constructor
}
```

In `async createStub`:
```js
if (!this.authClient && this.googleAuth) {
  // we are in async function, can await safely
  this.authClient = await this.googleAuth.getClient();
}

// then use this.authClient to get request headers
```

LibanOdowa commented:
Addressed changes, added the auth changes @alexander-fenster described.

One change needed in client libraries is for their webpack.config.js is to add a rule similar to the one I added for gax that ignores the google-auth-library. I will continue to check with browserify and see how that's possible in package.json in the future.
